client/src/controllers/PreviewController.ts

```diff
@@ -0,0 +1,426 @@
+  declare readonly hasModeSelectTarget: boolean;
+  declare readonly modeSelectTarget: HTMLSelectElement;
+  declare readonly iframeTarget: HTMLIFrameElement;
+  declare readonly iframeTargets: HTMLIFrameElement[];
+
+  declare readonly baseUrlValue: string;
+  declare readonly autoUpdateValue: boolean;
+  declare readonly autoUpdateIntervalValue: number;
+
+  // Instance variables with initial values set in connect()
+  declare editForm: HTMLFormElement;
+
+  // Instance variables with initial values set here
+  spinnerTimeout: ReturnType<typeof setTimeout> | null = null;
+  updatePromise: Promise<boolean> | null = null;
+  formPayload = '';
+
+  /**
+   * The default size input element.
+   * This is the size input element with the `data-default-size` data attribute.
+   * If no input element has this attribute, the first size input element will be used.
+   */
+  get defaultSizeInput(): HTMLInputElement {
+    return (
+      this.sizeInputTargets.find((input) => 'defaultSize' in input.dataset) ||
+      this.sizeInputTargets[0]
+    );
+  }
+
+  /**
+   * Sets the simulated device width of the preview iframe.
+   * @param width The width of the preview device. If falsy, the default size will be used.
+   */
+  setPreviewWidth(width?: string) {
+    const isUnavailable = this.element.classList.contains(
+      this.unavailableClass,
+    );
+
+    let deviceWidth = width;
+    // Reset to default size if width is falsy or preview is unavailable
+    if (!width || isUnavailable) {
+      deviceWidth = this.defaultSizeInput.dataset.deviceWidth;
+    }
+
+    this.element.style.setProperty(
+      '--preview-device-width',
+      deviceWidth as string,
+    );
+  }
+
+  /**
+   * Toggles the preview size based on the selected input.
+   * The selected device name (`input[value]`) is stored in localStorage.
+   * @param event `InputEvent` from the size input
+   */
+  togglePreviewSize(event: InputEvent) {
+    const target = event.target as HTMLInputElement;
+    const device = target.value;
+    const deviceWidth = target.dataset.deviceWidth;
+
+    this.setPreviewWidth(deviceWidth);
+    try {
+      localStorage.setItem('wagtail:preview-panel-device', device);
+    } catch (e) {
+      // Skip saving the device if localStorage fails.
+    }
+
+    // Ensure only one selected class is applied
+    this.sizeInputTargets.forEach((input) => {
+      // The <input> is invisible and we're using a <label> parent to style it.
+      (input.parentElement as HTMLLabelElement).classList.toggle(
+        this.selectedSizeClass,
+        input.value === device,
+      );
+    });
+  }
+
+  /**
+   * Observe the preview panel size and set the `--preview-panel-width` CSS variable.
+   * This is used to maintain the simulated device width as the side panel is resized.
+   */
+  observePanelSize() {
+    const resizeObserver = new ResizeObserver((entries) =>
+      this.element.style.setProperty(
+        '--preview-panel-width',
```

Comment:
Might be nice to have this variable name as a Stimulus value, just on that's always defaulted and not set in the html template. 

Especially as it's used in a few places.